lecture_07_idiomatic_python/materials/slides.ipynb

```diff
@@ -0,0 +1,2325 @@
+     "output_type": "stream",
+     "text": [
+      "this list contains so many strings \n"
+     ]
+    }
+   ],
+   "source": [
+    "strings = [\"this\", \"list\", \"contains\", \"so\", \"many\", \"strings\"]\n",
+    "s = \"\"\n",
+    "for string in strings:\n",
+    "    s += string + \" \"\n",
+    "print(s)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "This is job for **join**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "-"
+    }
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'this, list, contains, so, many, strings'"
+      ]
+     },
+     "execution_count": 20,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "\", \".join(strings)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'0 1 2 3 4 5 6 7 8 9'"
+      ]
+     },
+     "execution_count": 23,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "\" \".join(range(10))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Updating seuences"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "['new one', 'list', 'contains', 'so', 'many', 'strings']\n"
+     ]
+    }
+   ],
+   "source": [
+    "strings = [\"this\", \"list\", \"contains\", \"so\", \"many\", \"strings\"]\n",
+    "strings.pop(0)\n",
+    "strings.insert(0, \"new one\")\n",
+    "print(strings)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "deque(['new one', 'list', 'contains', 'so', 'many', 'strings'])\n"
+     ]
+    }
+   ],
+   "source": [
+    "from collections import deque\n",
+    "strings = deque([\"this\", \"list\", \"contains\", \"so\", \"many\", \"strings\"])\n",
+    "strings.popleft()\n",
+    "strings.appendleft(\"new one\")\n",
+    "print(strings)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1.47 ms ± 38.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)\n"
+     ]
+    }
+   ],
+   "source": [
+    "l = list(range(1_000_000))\n",
+    "%timeit l.pop(0); l.insert(0, 42)"
```

Comment:
Please also add pop right examples for the list and deque.   Both operations are O(1).